### PR TITLE
core[minor]: Adds an in-memory implementation of RecordManager

### PR DESCRIPTION
**Pull request by pprados**
**Description:**
langchain offers three technologies to save data:
- [vectorstore](https://python.langchain.com/docs/modules/data_connection/vectorstores/)
- [docstore](https://js.langchain.com/docs/api/schema/classes/Docstore)
- [record manager](https://python.langchain.com/docs/modules/data_connection/indexing)

If you want to combine these technologies in a sample persistence stategy you need a common implementation for each. `DocStore` propose `InMemoryDocstore`. 

We propose the class `MemoryRecordManager` to complete the system.

This is the prelude to another full-request, which needs a consistent combination of persistence components.

**Tag maintainer:**
@baskaryan

**Twitter handle:**
@pprados

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
this is good, but the async methods aren't actually async right?

**Comment by pprados** [comment]
No. All is async, because all method works in memories.

**Comment by pprados** [comment]
@hwchase17 
Some news ?

File: libs/community/langchain_community/indexes/memory_recordmanager.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    async def alist_keys(
+        self,
+        *,
+        before: Optional[float] = None,
+        after: Optional[float] = None,
+        group_ids: Optional[Sequence[str]] = None,
+        limit: Optional[int] = None,
+    ) -> List[str]:
+        return self.list_keys(
```
Comment (pprados) [inline libs/community/langchain_community/indexes/memory_recordmanager.py:113]: The code works only in memory. The async version is the same of the sync version

File: libs/community/langchain_community/indexes/memory_recordmanager.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            before=before, after=after, group_ids=group_ids, limit=limit
+        )
+
+    async def adelete_keys(self, keys: Sequence[str]) -> None:
+        return self.delete_keys(keys=keys)
```
Comment (pprados) [inline libs/community/langchain_community/indexes/memory_recordmanager.py:118]: The code works only in memory. The async version is the same of the sync version

File: libs/community/langchain_community/indexes/memory_recordmanager.py (hunk 1 above)
Comment (pprados) [inline libs/community/langchain_community/indexes/memory_recordmanager.py:113]: The code works only in memory. The async version is the same of the sync version

File: libs/community/langchain_community/indexes/memory_recordmanager.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            filter(
+                lambda record: record["namespace"] != self.namespace
+                or record["key"] not in keys,
+                self.data,
+            )
+        )
+
+    # %% Async versions
+    async def aget_time(self) -> float:
+        return datetime.now().timestamp()
```
Comment (pprados) [inline libs/community/langchain_community/indexes/memory_recordmanager.py:91]: The code works only in memory. The async version is the same of the sync version

File: libs/community/langchain_community/indexes/memory_recordmanager.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        self,
+        keys: Sequence[str],
+        *,
+        group_ids: Optional[Sequence[Optional[str]]] = None,
+        time_at_least: Optional[float] = None,
+    ) -> None:
+        return self.update(keys=keys, group_ids=group_ids, time_at_least=time_at_least)
+
+    async def aexists(self, keys: Sequence[str]) -> List[bool]:
+        return self.exists(keys=keys)
```
Comment (pprados) [inline libs/community/langchain_community/indexes/memory_recordmanager.py:103]: The code works only in memory. The async version is the same of the sync version

**Comment by pprados** [comment]
@baskaryan
I resynchronize the code for langchain-community version 0.0.14

**Comment by pprados** [comment]
@eyurtsev @hwchase17 
Hello,
Can you reconsider this PR?

**Comment by pprados** [comment]
@eyurtsev 
I rewrite the tests to be compatible with the same tests with SQLRecordManager.

**Comment by pprados** [comment]
@eyurtsev 
Can you check this ?

File: libs/community/langchain_community/indexes/_memory_recordmanager.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Sequence
+
+from langchain_community.indexes.base import RecordManager
+
+
+class MemoryRecordManager(RecordManager):
```
Comment (eyurtsev) [inline libs/community/langchain_community/indexes/_memory_recordmanager.py:7]: We have it in unit tests right now

https://github.com/langchain-ai/langchain/blob/main/libs/core/tests/unit_tests/indexing/in_memory.py#L13-L13

Do you want to just migrate it out of unit tests and into langchain_core?

I'm in favor of having InMemory implementations directly available in either langchain_core or langchain

Comment (pprados) [reply libs/community/langchain_community/indexes/_memory_recordmanager.py]: done

**Comment by pprados** [comment]
@eyurtsev 
I can result this bug.

with 
```
git checkout master
cd /libs/core
poetry lock --no-update
Resolving dependencies... (0.0s)

Because langchain-core depends on langchain-text-splitters (0.2.1) @ file:///home/pprados/workspace.bda/langchain/libs/text-splitters which depends on langchain-core (^0.2.0), langchain-core is required.
So, because langchain-core is 0.0.0a64, version solving failed.
```

**Comment by pprados** [comment]
@eyurtsev 
In the test CI /cd libs/core / make extended_tests, I receive an error:
```
INTERNALERROR>     raise Failed(msg=reason, pytrace=pytrace)
INTERNALERROR> Failed: Package `aiosqlite` is not installed but is required for extended tests. Please install the given package and try again.
```
It has nothing to do with my patch

**Comment by eyurtsev** [comment]
@pprados please review 

I reused the existing implementation of the in memory record manager from unit tests.
1. data is scoped to the instance not to the class (we generally don't want to scope data to the class namespace)
2. has slightly more documentation

It's moved to base.py together with the abstract interface. We can move it into a separate module as well (doesn't really matter), we just need it to be a non private module for it to appear in the API Reference documentation.

The typical name for memory implementations is "InMemory"

**Comment by pprados** [comment]
@eyurtsev 
It's perfect for me.

**Review by pprados — COMMENTED** [review]
Fixed

**Review by eyurtsev — APPROVED** [review]

**Pull request merged**